mozetl/taar/taar_ensemble.py

```diff
@@ -0,0 +1,525 @@
+    """
+
+    assert "ensemble_weights" in coefs
+    weights = coefs["ensemble_weights"]
+
+    assert len(weights) == 3
+
+    for key in weights.keys():
+        assert key in coefs["ensemble_weights"]
+        assert not np.isnan(coefs["ensemble_weights"][key])
+        assert coefs["ensemble_weights"][key] > 0.0
+
+    # This ordering must be strict
+    msg = """
+    FINAL WEIGHTS
+    =============
+    Collab     : {:0.8f}
+    Locale     : {:0.8f}
+    Similarity : {:0.8f}
+    """.format(
+        weights["collaborative"], weights["locale"], weights["similarity"]
+    )
+
+    print("Weight output")
+    print("================================")
+    print(msg)
+    print("================================")
+    assert weights["collaborative"] > 0.0
+    assert weights["locale"] > 0.0
+    assert weights["similarity"] > 0.0
+
+
+class CostLLR:
+    """ based on Niko Brummer's original implementation:
+        Niko Brummer and Johan du Preez, Application-Independent Evaluation of Speaker Detection"
+        Computer Speech and Language, 2005
+    """
+
+    def __init__(self):
+        self._total = 0
+
+    # evalcllr expects two lists
+    # recommendations_list should be a list of (guid, weight) 2-tuples
+    # unmasked_addons should be a list of guid strings
+    def evalcllr(self, recommendations_list, unmasked_addons):
+        # Organizer function to extract weights from recommendation list for passing to cllr.
+        lrs_on_target_helper = np.array(
+            [item[1] for item in recommendations_list if item[0] in unmasked_addons]
+        )
+        lrs_off_target_helper = np.array(
+            [item[1] for item in recommendations_list if item[0] not in unmasked_addons]
+        )
+        try:
+            tmp = self._cllr(lrs_on_target_helper, lrs_off_target_helper)
+        except Exception:
+            tmp = np.NaN
+
+        if np.isnan(tmp):
+            # This may happen if recommendations come back with a
+            # weight of 0
+            tmp = 0
+        self._total += tmp
+        return tmp
+
+    @property
+    def total(self):
+        return self._total
+
+    # Private methods below
+
+    # Helper function to do some math for cllr.
+    def _neg_log_sig(self, log_odds):
+        neg_log_odds = [-1.0 * x for x in log_odds]
+        e = np.exp(neg_log_odds)
+        return [np.log(1 + f) for f in e if f < (f + 1)]
+
+    # Compute the log likelihood ratio cost which should be minimized.
+    def _cllr(self, lrs_on_target, lrs_off_target):
+        lrs_on_target = np.log(lrs_on_target[~np.isnan(lrs_on_target)])
+        lrs_off_target = np.log(lrs_off_target[~np.isnan(lrs_off_target)])
+
+        c1 = np.mean(self._neg_log_sig(lrs_on_target)) / np.log(2)
+        c2 = np.mean(self._neg_log_sig(-1.0 * lrs_off_target)) / np.log(2)
+        return (c1 + c2) / 2
+
+
+def cross_validation_split(dataset, n_folds):
+    """
+  Splits dataframe into n_folds, returning array of dataframes
```

Comment:
This could be renamed to `k_folds` to be consistent with other definitions of cross-validation.

Reply:
Fixed in 6e0befa